mathesar/models.py

```diff
@@ -16,32 +19,56 @@ class Meta:
 
 
 class DatabaseObject(BaseModel):
-    name = models.CharField(max_length=63)
+    oid = models.IntegerField()
 
     class Meta:
         abstract = True
 
     def __str__(self):
-        return f"{self.__class__.__name__}: {self.name}"
+        return f"{self.__class__.__name__}: {self.oid}"
 
 
 class Schema(DatabaseObject):
     database = models.CharField(max_length=128)
 
+    @cached_property
+    def _sa_engine(self):
+        # We're caching this since the engine is used frequently.
+        return create_mathesar_engine(self.database)
+
+    @cached_property
+    def name(self):
+        cache_key = f"{self.database}_schema_name_{self.oid}"
+        try:
+            schema_name = cache.get(cache_key)
+            if schema_name is None:
+                schema_name = schemas.get_schema_name_from_oid(
+                    self.oid, self._sa_engine
+                )
+                cache.set(cache_key, schema_name, NAME_CACHE_INTERVAL)
+            return schema_name
+        except TypeError:
+            return 'MISSING'
+
 
 class Table(DatabaseObject):
     schema = models.ForeignKey('Schema', on_delete=models.CASCADE,
                                related_name='tables')
     import_verified = models.BooleanField(blank=True, null=True)
 
     @cached_property
-    def _sa_engine(self):
-        # We're caching this since the engine is used frequently.
-        return create_mathesar_engine(self.schema.database)
-
-    @property
     def _sa_table(self):
-        return tables.reflect_table(self.name, self.schema.name, self._sa_engine)
+        try:
+            table = tables.reflect_table_from_oid(
+                self.oid, self.schema._sa_engine,
+            )
+        except TypeError:
+            table = tables.create_empty_table("MISSING")
```

Comment:
If there's no table matching the `oid` in the database, we should delete the `Table` object in the Django DB, since we should be reflecting the tables accurately.

Same with schemas.

Reply:
I see below that we are doing this, so I'm not sure why we're creating an empty table here. Can you add a comment to explain?

Reply:
This is to handle the case where the user is looking at a specific table, and it's deleted under their feet.  Basically, on whatever the next API call is, the front end would get back a table named `MISSING`, and could (for example) just pop up with "the table was deleted" rather than "table not found", before redirecting back to the list.  Once it's redirected back to the list, the DB Sync function will kick in and delete the table's record.

The side benefit is that it gives us the flexibility to decouple the cadence of the overall DB Syncing from the expiration of the cache with the table's name.

I'll put a shorter version in a code comment.

Reply:
That makes sense. I'll merge it for now, we might want to figure out a better way to handle this case in the future (there might actually be empty tables named `MISSING` if the user creates them).

Reply:
Agreed.  Another option I considered was to just raise a custom exception, and then pass that via the API to the front end.

Reply:
I like that idea better.